flang/lib/Frontend/FrontendActions.cpp

```diff
@@ -709,64 +709,81 @@ void CodeGenAction::lowerHLFIRToFIR() {
-          << outputErrMsg.str();
-      return std::nullopt;
-    }
+  auto parseResult = llvm::RISCVISAInfo::parseFeatures(
+      triple.isRISCV64() ? 64 : 32, targetOpts.featuresAsWritten);
+  if (!parseResult) {
+    std::string buffer;
+    llvm::raw_string_ostream outputErrMsg(buffer);
+    handleAllErrors(parseResult.takeError(), [&](llvm::StringError &errMsg) {
+      outputErrMsg << errMsg.getMessage();
+    });
+    ci.getDiagnostics().Report(clang::diag::err_invalid_feature_combination)
+        << outputErrMsg.str();
+    return std::nullopt;
+  }
 
-    llvm::RISCVISAInfo *const isaInfo = parseResult->get();
+  llvm::RISCVISAInfo *const isaInfo = parseResult->get();
 
-    // RISCV::RVVBitsPerBlock is 64.
-    unsigned vscaleMin = isaInfo->getMinVLen() / llvm::RISCV::RVVBitsPerBlock;
+  // RISCV::RVVBitsPerBlock is 64.
+  unsigned vscaleMin = isaInfo->getMinVLen() / llvm::RISCV::RVVBitsPerBlock;
 
-    if (langOpts.VScaleMin || langOpts.VScaleMax) {
-      // Treat Zvl*b as a lower bound on vscale.
-      vscaleMin = std::max(vscaleMin, langOpts.VScaleMin);
-      unsigned vscaleMax = langOpts.VScaleMax;
-      if (vscaleMax != 0 && vscaleMax < vscaleMin)
-        vscaleMax = vscaleMin;
-      return std::pair<unsigned, unsigned>(vscaleMin ? vscaleMin : 1,
-                                           vscaleMax);
-    }
+  if (langOpts.VScaleMin || langOpts.VScaleMax) {
+    // Treat Zvl*b as a lower bound on vscale.
+    vscaleMin = std::max(vscaleMin, langOpts.VScaleMin);
+    unsigned vscaleMax = langOpts.VScaleMax;
+    if (vscaleMax != 0 && vscaleMax < vscaleMin)
+      vscaleMax = vscaleMin;
+    return std::pair<unsigned, unsigned>(vscaleMin ? vscaleMin : 1, vscaleMax);
+  }
 
-    if (vscaleMin > 0) {
-      unsigned vscaleMax = isaInfo->getMaxVLen() / llvm::RISCV::RVVBitsPerBlock;
-      return std::make_pair(vscaleMin, vscaleMax);
-    }
-  } else {
-    if (langOpts.VScaleMin || langOpts.VScaleMax)
-      return std::pair<unsigned, unsigned>(
-          langOpts.VScaleMin ? langOpts.VScaleMin : 1, langOpts.VScaleMax);
+  if (vscaleMin > 0) {
+    unsigned vscaleMax = isaInfo->getMaxVLen() / llvm::RISCV::RVVBitsPerBlock;
+    return std::make_pair(vscaleMin, vscaleMax);
   }
 
   return std::nullopt;
 }
 
+// TODO: We should get this from TargetInfo. However, that depends on
+// too much of clang, so for now, replicate the functionality.
+static std::optional<std::pair<unsigned, unsigned>>
+getVScaleRange(CompilerInstance &ci) {
+  const auto &langOpts = ci.getInvocation().getLangOpts();
+  const llvm::Triple triple(ci.getInvocation().getTargetOpts().triple);
+
+  if (triple.isAArch64())
+    return getAArch64VScaleRange(ci);
+  if (triple.isRISCV())
+    return getRISCVVScaleRange(ci);
+
+  if (langOpts.VScaleMin || langOpts.VScaleMax)
+    return std::pair<unsigned, unsigned>(
+        langOpts.VScaleMin ? langOpts.VScaleMin : 1, langOpts.VScaleMax);
```

Comment:
It will execute on anything that's not AArch64 or RISC-V. I'm not sure if it will have any effect on codegen though. I kept this here to preserve the behaviour before this patch, but happy to remove it here if you'd like

Reply:
This would make more sense to me:
```
getVScaleRange(CompilerInstance &ci) {
  if (AArch64)
    return do_something;
  if (RISC-V)
    return do_something_else;
  
  // All other architectures that don't support scalable vectors (i.e. don't need vscale)
  return nullopt;
}
```

Reply:
I specifically object to this suggestion.

We had exactly this code structure previously, and it broke RISCV because the author only considered AArch64.  Why should we bother to do something to actively break other (out of tree?) targets which might support scalable vectors?

Reply:
> We had exactly this code structure previously, and it broke RISCV 

We are obviously very excited to see all the RISC-V contributions for Flang - it's great to see the increasing adoption! However, we need to be realistic - Flang is a very young project and the support for various architectures depends heavily on where the most contributions come from. A good proxy for which platforms are supported _well_ can be found in two places:
* https://github.com/llvm/llvm-project/blob/main/flang/docs/GettingStarted.md#supported-c-compilers
* https://lab.llvm.org/buildbot/#/builders (type "flang" in the search box).

RISC-V is not listed there - my assumption is that it reflects the fact that the interest is still rather limited.

Importantly, we can't really reason about the status of RISC-V support in Flang - we have no data for that. RISC-V buildbot for Flang would greatly help with patches like this one.

> Why should we bother to do something to actively break other (out of tree?) targets 

That's not the intent here and I don't know where you're getting that from. Instead, we are trying to make the level of the available support very clear. 

The corresponding options are only intended for AArch64 and RISC-V:

* https://github.com/llvm/llvm-project/blob/114e6d7ba02f090117f2cb1ffeb9027cf80f335b/clang/include/clang/Driver/Options.td#L4695-L4703

The driver should enforce that and that's what I propose here:

* https://github.com/llvm/llvm-project/pull/77905

If the preference is to indeed support other targets that implement scalable vectors (and require these options) then that should be tested. My preference is to simply limit the usage of `-mvscale-max` and `-mvscale-min` to AArch64+RISC-V.

Reply:
@banach-space You seem to have taken my comment in an unexpected direction.  I was not arguing that RISCV was "supported" in any particular way.  I was making a purely stylistic point that intentionally breaking other targets seems undesirable.  In order for RISCV (or any target) to become supported, we have to do bringup work.  This patch, and the others for RISCV, are specifically around getting RISCV support to a point where it could be reasonable called supported.  We're not there yet; I don't think anyone would say otherwise.  

My argument for the current code structure is that supporting other future targets with reasonable default code structure will make the bringup efforts for them more straight forward.  As evidence for my argument, I used the fact that the prior code structure did exactly that for RISCV.  This is a stylistic concern, and I'll fully admit one we could reasonable disagree on.  I think you're wrong here, but if you care strongly about this particular stylistic question, I will defer to you because I just don't think it matters that much.

Reply:
@banach-space The direction in #77905 makes sense to me, thanks. To go forward, is this OK to land as is, and then in a separate commit/PR we would make the else branch here a llvm_unreachable?

Reply:
> @banach-space The direction in #77905 makes sense to me, thanks. To go forward, is this OK to land as is, and then in a separate commit/PR we would make the else branch here a llvm_unreachable?

This makes a lot of sense to me, thank you both for the discussion!

Reply:
> @banach-space You seem to have taken my comment in an unexpected direction. I was not arguing that RISCV was "supported" in any particular way. 

I was referring to this comment specifically:

> We had exactly this code structure previously, and it broke RISCV because the author only considered AArch64. 

I was trying to say that:
* IMHO, in this case, RISC-V  was "not supported yet" rather than "broken",
* the previous structure strove to communicate that, i.e. that only `AArch64` was supported (the code could've been more explicit about it - perhaps with `llvm_unreachable`?).

That kind of situation is quite common within Flang - many targets/features are just not supported yet. Also, in this case, I wouldn't expect somebody working on `AArch64` to know the semantics of `-mvscale-min/-mvscale-max` on RISC-V (and vice-versa).

I've probably read your comment too literally, but also want to manage expectations. And encourage you to consider adding public buildbots - that's has been incredibly helpful in making sure we have strong `AArch64` support.

Thanks for contributing 🙏🏻 !